src/Maestro/Maestro.GitHub/GitHubTokenProvider.cs

```diff
@@ -52,5 +62,39 @@ private string GetAppToken()
                 });
             return generator.CreateEncodedJwtToken();
         }
+
+        private bool TryGetCachedToken(long installationId, out string cachedToken)
+        {
+            cachedToken = null;
+
+            if (!_tokenCache.ContainsKey(installationId))
+            {
+                return false;
+            }
+
+            AccessToken token = _tokenCache[installationId];
+
+            // If the cached token will expire in less than 15 minutes we won't use it and let GetTokenForInstallation generate a new one
+            // and update the cache
+            if (DateTimeOffset.Now.Subtract(token.ExpiresAt).TotalMinutes < 15)
+            {
+                return false;
+            }
+
+            cachedToken = token.Token;
+            return true;
+        }
+
+        private void UpdateTokenCache(long installationId, AccessToken accessToken)
+        {
+            if (_tokenCache.ContainsKey(installationId))
```

Comment:
Does this code need to be thread-safe? I think the singleton's are per-service, so I'm guessing you can call this code simultaneously from multiple actors.

Reply:
Yes, this code needs to be thread safe.